File: C++/myLeetCode/Dynamic_Programming/Subsequence_problem/115. Distinct Subsequences.cpp
// 115. Distinct Subsequences
// https://leetcode.com/problems/distinct-subsequences/

#include <iostream>
#include <string>
#include <vector>

int numDistinct(std::string s, std::string t) {
    int sLen = s.length();
    int tLen = t.length();

    std::vector<std::vector<long long>> dp(sLen + 1, std::vector<long long>(tLen + 1, 0));

    for (int i = 0; i <= sLen; i++) {
        dp[i][0] = 1;
    }

    for (int i = 1; i <= sLen; i++) {
        for (int j = 1; j <= tLen; j++) {
            if (s[i - 1] == t[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j];
            } else {
                dp[i][j] = dp[i - 1][j];
            }
        }
    }

    return dp[sLen][tLen];
}

int main() {
    std::vector<std::pair<std::string, std::string>> testCases = {
        {"rabbbit", "rabbit"},
        {"babgbag", "bag"}
    };

    for (int i = 0; i < testCases.size(); i++) {
        std::string s = testCases[i].first;
        std::string t = testCases[i].second;

        int num = numDistinct(s, t);

        std::cout << "Number of distinct subsequences for test case " << (i + 1) << ": ";
        std::cout << num << std::endl;
    }

    return 0;
}

File: C++/代码随想录/Dynamic_Programming/Subsequence_problem/115. Distinct Subsequences.py
def numDistinct(s, t):
    m, n = len(s), len(t)
    if m < n:
        return 0

    # dp[i][j] 表示 s[:i] 中有多少个子序列等于 t[:j]
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    # 当 t 为空字符串时，s 的每个子串都有一个空子序列与之匹配
    for i in range(m + 1):
        dp[i][0] = 1

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s[i - 1] == t[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j]
            else:
                dp[i][j] = dp[i - 1][j]

    return dp[m][n]

# 测试用例
print(numDistinct("rabbbit", "rabbit"))  # 输出应该是 3
